### PR TITLE
[Modelzoo] Rebuild DeepFM to update API and enable DeepRec features.

### PR DESCRIPTION
**Pull request by chenzhiyin**
TODO:

enable following DeepRec feature:
7) StarServer
**_8) Auto Micro Batch same with https://github.com/alibaba/DeepRec/issues/127_**
9) FusedEmbedding API waiting related PR merged.
**_10) Smart Stage same with https://github.com/alibaba/DeepRec/issues/122
11) Auto Graph Fusion https://github.com/alibaba/DeepRec/issues/144_**
12) CPU Memory Optimization:START_STATISTIC_STEP, STOP_STATISTIC_STEP, jemalloc
14) AdamAsync Optimizer
15) BF16

Feature Doc:
1 ) Embedding Variable
7) GRPC++
13) Incremental Checkpoint
14) AdagradDecay

（optional, default to closed.)
2) EmbeddingVariable进阶功能：特征淘汰
3) EmbeddingVariable进阶功能：特征准入
**_4) Dynamic-dimension Embedding Variable_**
5) Adaptive Embedding
**_6) Multi-Hash Variable https://github.com/alibaba/DeepRec/issues/145_**
17) WorkQueue

### CODE REVIEW

File: modelzoo/DeepFM/train.py [hunk 1]
```diff
@@ -14,6 +14,7 @@
 tf.logging.set_verbosity(tf.logging.INFO)
 print("Using TensorFlow version %s" % (tf.__version__))
 
+# Definition of some constants
 CONTINUOUS_COLUMNS = ["I" + str(i) for i in range(1, 14)]  # 1-13 inclusive
 CATEGORICAL_COLUMNS = ["C" + str(i) for i in range(1, 27)]  # 1-26 inclusive
 LABEL_COLUMN = ["clicked"]
```
Comment (Duyi-Wang) [inline modelzoo/DeepFM/train.py:20]: Use single quotes instead of double quotes

File: modelzoo/DeepFM/train.py [hunk 2]
```diff
@@ -579,7 +819,7 @@ def main(tf_config=None, server=None):
                     'index': task_index,
                     'is_chief': is_chief
                 },
-                     server=server)
+                    server=server)
```
Comment (Duyi-Wang) [inline modelzoo/DeepFM/train.py:822]: Please use `generate_cluster_info()` to Parse distributed training configuration and generate cluster information

File: modelzoo/DeepFM/train.py [hunk 3]
```diff
@@ -406,119 +530,235 @@ def main(tf_config=None, server=None):
...
+                        type=int,
+                        default=0)
+    parser.add_argument('--input_layer_partitioner',
+                        help='slice size of input layer partitioner, units MB. Default 8MB',
+                        type=int,
+                        default=0)
+    parser.add_argument('--dense_layer_partitioner',
+                        help='slice size of dense layer partitioner, units KB. Default 16KB',
+                        type=int,
+                        default=0)
```
Comment (Duyi-Wang) [inline modelzoo/DeepFM/train.py:630]: Set right default value for `layer_partitioner` params

File: modelzoo/DeepFM/train.py [hunk 4]
```diff
@@ -406,119 +530,235 @@ def main(tf_config=None, server=None):
...
+    parser.add_argument('--seed',
+                        help='set the random seed for tensorflow',
+                        type=int,
+                        default=2021)
+    parser.add_argument("--optimizer",
+                        type=str,
+                        choices=['adam', 'adamasync',
+                                 'adagraddecay', 'adagrad'],
+                        default='adam')
```
Comment (Duyi-Wang) [inline modelzoo/DeepFM/train.py:598]: Set default to `adamasync`

File: modelzoo/DeepFM/train.py [hunk 5]
```diff
@@ -406,119 +530,235 @@ def main(tf_config=None, server=None):
...
+    # Run model training and evaluation
+    train(sess_config, hooks, model, train_init_op, train_steps,
+          checkpoint_dir, tf_config, server)
+    if not (args.no_eval or tf_config):
+        eval(sess_config, hooks, model, test_init_op, test_steps,
+             checkpoint_dir)
+
+# Get parse
+
+
```
Comment (Duyi-Wang) [inline modelzoo/DeepFM/train.py:562]: remove blank line

File: modelzoo/DeepFM/train.py [hunk 6]
```diff
@@ -406,119 +530,235 @@ def main(tf_config=None, server=None):
                    input_layer_partitioner=input_layer_partitioner,
                    dense_layer_partitioner=dense_layer_partitioner)
```
Comment (Duyi-Wang) [inline modelzoo/DeepFM/train.py:532]: stock_tf and adaptive_emb missed

File: modelzoo/DeepFM/train.py [hunk 7]
```diff
@@ -210,122 +171,287 @@ def prediction(self):
...
+        1539.0, 22069.0, 65535.0, 561.0, 2655388.0, 233523.0, 26297.0, 5106.0,
+        24376.0, 9.0, 181.0, 1807.0, 6879.0
+    ]
+
+    def make_minmaxscaler(min, range):
+        def minmaxscaler(col):
+            return (col - min) / range
+
+        return minmaxscaler
+
```
Comment (Duyi-Wang) [inline modelzoo/DeepFM/train.py:286]: remove normalizer. Follow DeepFM code.

File: modelzoo/DeepFM/train.py [hunk 8]
```diff
@@ -210,122 +171,287 @@ def prediction(self):
...
+                learning_rate=self._learning_rate,
                 beta1=0.9,
                 beta2=0.999,
                 epsilon=1e-8)
+        elif self._optimizer_type == 'adagraddecay':
+            optimizer = tf.train.AdagradDecayOptimizer(
+                learning_rate=self._learning_rate,
+                global_step=self.global_step)
         else:
             raise ValueError('Optimzier type error.')
```
Comment (Duyi-Wang) [inline modelzoo/DeepFM/train.py:217]: Set to adam if self.tf

File: modelzoo/DeepFM/train.py (hunk 4 above)
Comment (chenzhiyin) [reply modelzoo/DeepFM/train.py]: deafult value is set as "adamasync" now

File: modelzoo/DeepFM/train.py (hunk 5 above)
Comment (chenzhiyin) [reply modelzoo/DeepFM/train.py:564]: they are removed

File: modelzoo/DeepFM/train.py (hunk 6 above)
Comment (chenzhiyin) [reply modelzoo/DeepFM/train.py:529]: they are added

File: modelzoo/DeepFM/train.py (hunk 7 above)
Comment (chenzhiyin) [reply modelzoo/DeepFM/train.py:285]: it is removed

File: modelzoo/DeepFM/train.py (hunk 8 above)
Comment (chenzhiyin) [reply modelzoo/DeepFM/train.py:216]: revised

File: modelzoo/DeepFM/train.py (hunk 1 above)
Comment (chenzhiyin) [reply modelzoo/DeepFM/train.py]: revised

File: modelzoo/DeepFM/train.py (hunk 7 above)
Comment (Duyi-Wang) [reply modelzoo/DeepFM/train.py:285]: mins_list, range_list and make_minmaxscaler() is still here

File: modelzoo/DeepFM/train.py [hunk 9]
```diff
@@ -402,184 +521,259 @@ def main(tf_config=None, server=None):
...
+    More Detail information is shown in https://deeprec.readthedocs.io/zh/latest/index.html.
+    START_STATISTIC_STEP & STOP_STATISTIC_STEP: On CPU platform, DeepRec supports memory optimization
+        in both stand-alone and distributed trainging. It's default to open, and the 
+        default start and stop steps of collection is 1000 and 1100. Reduce the initial 
+        cold start time by the following settings.
+    MALLOC_CONF: On CPU platform, DeepRec can use memory optimization with the jemalloc library.
+        Please preload libjemalloc.so by `LD_PRELOAD=./libjemalloc.so.2 python ...`
+    '''
+    os.environ['START_STATISTIC_STEP'] = '100'
+    os.environ['STOP_STATISTIC_STEP'] = '200'
```
Comment (Duyi-Wang) [inline modelzoo/DeepFM/train.py:762]: Set 'STOP_STATISTIC_STEP' to 110. We get some update on this.

File: modelzoo/DeepFM/train.py [hunk 10]
```diff
@@ -402,184 +521,259 @@ def main(tf_config=None, server=None):
...
+                        default=None)
+    parser.add_argument('--ev_filter',
+                        help='Feature Filter of EmbeddingVariable Feature. Default closed.',
+                        type=str,
+                        choices=[None, 'counter', 'cbf'],
+                        default=None)
+    parser.add_argument('--op_fusion',
+                        help='Whether to enable Auto graph fusion feature. Default to True',
+                        type=boolean_string,
+                        default=True)  # TODO: Default to be True
```
Comment (Duyi-Wang) [inline modelzoo/DeepFM/train.py:668]: Please remove TODO if done.

File: modelzoo/DeepFM/train.py [hunk 11]
```diff
@@ -402,184 +506,259 @@ def main(tf_config=None, server=None):
...
-                # save_checkpoint_steps=args.save_steps,
-                log_step_count_steps=100,
-                config=sess_config) as sess:
-            while not sess.should_stop():
-                _, train_loss = sess.run([model.train_op, model.loss])
+    if args.smartstaged and not args.tf:
+        '''Smart staged Feature'''
+        next_element = tf.staged(next_element, num_threads=8, capacity=40)
+        sess_config.graph_options.optimizer_options.do_smart_stage = True
+        hooks.append(tf.make_prefetch_hook())
```
Comment (Duyi-Wang) [inline modelzoo/DeepFM/train.py:527]: Refer to WDL to enable smartstaged feature

File: modelzoo/DeepFM/train.py [hunk 12]
```diff
@@ -210,122 +171,267 @@ def prediction(self):
...
-            net = self.dnn(all_input, self.final_hidden_units, 'final_dnn')
+                else final_dnn_scope:
+            net = self._dnn(all_input, self._final_hidden_units, 'final_dnn')
 
         if self.bf16:
             net = tf.cast(net, dtype=tf.float32)
 
         net = tf.layers.dense(net, units=1)
-        net = tf.math.sigmoid(net)
+        self.predict = tf.math.sigmoid(net)
```
Comment (Duyi-Wang) [inline modelzoo/DeepFM/train.py:181]: self._logits

Comment (app-on-mic) [reply modelzoo/DeepFM/train.py:181]: @chenzhiyin  Since our customer is waiting for the update, please help to modify the code.  If any help is needed, pls let us know. Thanks.

Comment (chenzhiyin) [reply modelzoo/DeepFM/train.py:181]: OK, I am working on it.

**Review by chenzhiyin — COMMENTED** [review]
Thank the comments of Duyi, I have revised the corresponding codes.

File: modelzoo/DeepFM/train.py [hunk 13]
```diff
@@ -395,191 +497,266 @@ def main(tf_config=None, server=None):
...
+                        choices=[None, 'counter', 'cbf'],
+                        default=None)
+    parser.add_argument('--op_fusion',
+                        help='Whether to enable Auto graph fusion feature. Default to True',
+                        type=boolean_string,
+                        default=True)
+    parser.add_argument('--micro_batch',
+                        help='Set num for Auto Mirco Batch. Default close.',
+                        type=int,
+                        default=0)
```
Comment (Duyi-Wang) [inline modelzoo/DeepFM/train.py:655]: ADD #TODO flag, because it isn't really enabled. When Ali fixes the related issue, we need to revise the code.

File: modelzoo/DeepFM/train.py [hunk 14]
```diff
@@ -395,191 +497,266 @@ def main(tf_config=None, server=None):
...
+    parser.add_argument('--adaptive_emb',
+                        help='Whether to enable Adaptive Embedding. Default to False.',
+                        type=boolean_string,
+                        default=False)
+    parser.add_argument('--dynamic_ev',
+                        help='Whether to enable Dynamic-dimension Embedding Variable. Default to False.',
+                        type=boolean_string,
+                        default=False)
```
Comment (Duyi-Wang) [inline modelzoo/DeepFM/train.py:663]: ADD #TODO flag, because it isn't really enabled. When Ali fixes the related issue, we need to revise the code.

File: modelzoo/DeepFM/train.py [hunk 15]
```diff
@@ -210,149 +171,286 @@ def prediction(self):
...
+                    '''
+                    categorical_column = tf.feature_column.categorical_column_with_adaptive_embedding(
+                        column_name,
+                        hash_bucket_size=HASH_BUCKET_SIZES[column_name],
+                        dtype=tf.string,
+                        ev_option=ev_opt)
+                elif args.dynamic_ev:
+                    '''Dynamic-dimension Embedding Variable'''
+                    print(
+                        "Dynamic-dimension Embedding Variable isn't real enabled in model."
```
Comment (Duyi-Wang) [inline modelzoo/DeepFM/train.py:325]: really enabled

File: modelzoo/DeepFM/train.py [hunk 16]
```diff
@@ -210,149 +171,286 @@ def prediction(self):
...
+    save_steps = args.save_steps if args.save_steps or args.no_eval else steps
+    '''
+                            Incremental_Checkpoint
+    Please add `save_incremental_checkpoint_secs` in 'tf.train.MonitoredTrainingSession'
+    it's default to None, Incremental_save checkpoint time in seconds can be set 
+    to use incremental checkpoint function, like `tf.train.MonitoredTrainingSession(
+        save_incremental_checkpoint_secs=args.incremental_ckpt)`
+    '''
+    if args.incremental_ckpt and not args.tf:
+        print("Incremental_Checkpoint is not real enabled.")
```
Comment (Duyi-Wang) [inline modelzoo/DeepFM/train.py:390]: really enabled
